### ISSUE TITLE
Need guidance on how you have implemented the reconstruction loss 

### ISSUE THREAD

**Issue by ksachdeva** [sebasutp/trajectory_forcasting#1]
Hi,

Thanks for a very interesting and insightful paper as well very clean code.

I believe I have understood most of the key insights & able to follow the code except for the "second term" of the loss function.

```python
        # Compute Exp Likelihood
        sig_y = keras.backend.exp(log_sig_y)
        d = y - y_decoded_mean
        d_sq = keras.backend.square(d)
        d_mah = tf.divide(d_sq, sig_y)
        log_det_sig_y = keras.backend.sum(log_sig_y) #has to be a scalar
        d_mah_sum = keras.backend.sum(d_mah, axis=-1 ) + log_det_sig_y
        d_sq_masked = d_mah_sum * mask

        rec_loss = 0.5 * keras.backend.sum( d_sq_masked )
```

From your paper, 

![image](https://user-images.githubusercontent.com/127006/88362050-c1212f80-cd40-11ea-8369-a4d88f8bfe69.png)

The highlighted part seems to be about the second term (reconstruction loss) and mentions that expectation is approximated by Monte Carlo sampling. I am not able to connect the theory with the code snippet mentioned above ....especially following lines

```
d_mah = tf.divide(d_sq, sig_y)
log_det_sig_y = keras.backend.sum(log_sig_y) #has to be a scalar
d_mah_sum = keras.backend.sum(d_mah, axis=-1 ) + log_det_sig_y 
```

And not able to see the monte carlo sampling aspect either for this second term.

Would really appreciate if you could guide

Regards & thanks
Kapil

**Comment by ksachdeva** [comment]
Hi,

After mulling over these lines more and other parts I noticed that I had a misunderstanding regarding the output of your decoder.

I now see that your decoder is outputting `mu` & `log_sigma`  as well. Since most of the decoders directly output predictions my mind got anchored by that fact.

Using this I can now understand that these lines 

```
        # Compute Exp Likelihood
        sig_y = keras.backend.exp(log_sig_y)
        d = y - y_decoded_mean
        d_sq = keras.backend.square(d)
        d_mah = tf.divide(d_sq, sig_y)
        log_det_sig_y = keras.backend.sum(log_sig_y) #has to be a scalar
        d_mah_sum = keras.backend.sum(d_mah, axis=-1 ) + log_det_sig_y
        d_sq_masked = d_mah_sum * mask

```

are simply computing the log-likelihood assuming a normal distribution. `y_decoded_mean` is the `mu` outputted by the decoder and `log_sig_y` is also output from your decoder; it's just that you have added this trainable variable in a bit unorthodox manner. 

I really liked your paper and the interesting use of CVAE. Thanks again.

Regards
Kapil
